### ISSUE TITLE
Fullscreen windows minimized on focus loss don't restore correctly in Wine

### ISSUE THREAD

**Issue by madewokherd** [libsdl-org/SDL#5320]
In SDL 2.0.18 on Wine, if a fullscreen window loses focus and is minimized, it's no longer restored fullscreen.

I don't think this technically a regression, as previous versions did not notice the focus loss at all. So minimization would've always caused this problem, it just never happened before. When SDL minimizes its own window, it unsets fullscreen first and, in Wine specifically, it fails to set it to fullscreen on restore.

I haven't been able to figure out exactly why this is. SDL definitely requests that its window be resized/moved to fullscreen. Wine responds by requesting a resize/move and requesting NET_WM_STATE_FULLSCREEN. Yet for some reason, afterwards we get a ConfigureNotify with the old size/position, and the window is never changed back to fullscreen. So it looks like a window manager bug to me, but I don't know what we could be doing to cause this behavior.

Test case:
```c
#include <SDL2/SDL.h>
#include <stdio.h>

int wmain(int argc, wchar_t* argv[])
{
    SDL_Window *window;
    int is_fullscreen = 0;

    SDL_Init(SDL_INIT_EVERYTHING);

    SDL_SetHint(SDL_HINT_VIDEO_MINIMIZE_ON_FOCUS_LOSS, "1");

    window = SDL_CreateWindow("Test", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, 320, 240, 0);

    if (!window)
        return 1;

    while (1)
    {
        SDL_Event event;

        if (SDL_WaitEvent(&event))
        {
            switch (event.type)
            {
            case SDL_QUIT:
                printf("got SDL_QUIT\n");
                SDL_DestroyWindow(window);
                SDL_Quit();
                return 0;
            case SDL_MOUSEBUTTONDOWN:
                is_fullscreen = !is_fullscreen;
                printf("Setting fullscreen: %i\n", is_fullscreen);
                SDL_SetWindowFullscreen(window, is_fullscreen ? SDL_WINDOW_FULLSCREEN_DESKTOP : 0);
                printf("Done setting fullscreen\n");
                break;
            default:
                break;
            }
        }
        else
        {
            printf("SDL_WaitEvent failed %s\n", SDL_GetError());
        }
    }

    return 0;
}

int main(int argc, char* argv[])
{
    return wmain(0, NULL);
}
```

To reproduce the bug:
 * Run the test program in Wine.
 * Click the window to fullscreen.
 * Alt+tab to switch away.
 * Switch back.
 * Window is borderless 320x240, rather than fullscreen.

This doesn't happen on X11, even though SDL should interact with the window manager in the same way, and the window is not fullscreen when minimized. I don't know why.

**Comment by madewokherd** [comment]
IMO, the "unfullscreen on minimize" behavior is wrong on most platforms, even though it seems to work. On both Windows and X11, a window can be fullscreen and minimized at the same time. Both platforms also can preview the contents of minimized windows, which isn't going to look right if the window is resized first. We do still need the mode change, and on Windows we may still need to update the size on restore, if it's FULLSCREEN_DESKTOP and the desktop was resized.

I can see there is a fair amount of macosx-specific logic, so I guess it is needed there at least. I'd guess that many of the other platforms don't have a concept of fullscreen or minimization.

**Comment by madewokherd** [comment]
My window manager is Xfwm4, but I don't know if it's WM-specific.

**Comment by slouken** [comment]
@icculus, this is affecting games under Proton, can you take a look?

**Comment by madewokherd** [comment]
Well, technically it isn't yet, to my knowledge, because it requires a game using the minimize hint to upgrade to one of these versions. We found it from trying to update the SDL in Wine Mono, which broke One Finger Death Punch and DLC Quest.

**Comment by icculus** [comment]
I can look into this, but to be clear: if this works in Windows and not Wine, isn't this a Wine bug and not an SDL bug?

**Comment by madewokherd** [comment]
It's likely either a Wine bug or a WM bug. AFAICT SDL isn't doing anything wrong, just something unnecessary that happens to cause problems in Wine, for a reason I don't fully understand yet.

**Comment by slouken** [comment]
> I can look into this, but to be clear: if this works in Windows and not Wine, isn't this a Wine bug and not an SDL bug?

Yep, it's probably a Wine/Proton bug, and we can upstream any patch needed to fix this.

**Comment by madewokherd** [comment]
Well, this is what I came up with as a work-around:
```diff
diff --git a/src/video/windows/SDL_windowswindow.c b/src/video/windows/SDL_windowswindow.c
index 53365f160..b84c9e0e8 100644
--- a/src/video/windows/SDL_windowswindow.c
+++ b/src/video/windows/SDL_windowswindow.c
@@ -688,6 +688,13 @@ WIN_SetWindowFullscreen(_THIS, SDL_Window * window, SDL_VideoDisplay * display,
     int x, y;
     int w, h;
 
+	if (!fullscreen && (window->flags & (SDL_WINDOW_FULLSCREEN_DESKTOP|SDL_WINDOW_FULLSCREEN)) != 0)
+	{
+		/* Resizing the window on hide causes problems restoring it in Wine, and it's unnecessary.
+		 * Also, Windows would preview the minimized window with the wrong size. */
+		return;
+	}
+
     if (SDL_ShouldAllowTopmost() && ((window->flags & (SDL_WINDOW_FULLSCREEN|SDL_WINDOW_INPUT_FOCUS)) == (SDL_WINDOW_FULLSCREEN|SDL_WINDOW_INPUT_FOCUS) || window->flags & SDL_WINDOW_ALWAYS_ON_TOP)) {
         top = HWND_TOPMOST;
     } else {
```

Basically, don't unfullscreen the window on Windows if it's logically still in fullscreen mode. I don't expect this to break anything on either Windows or Wine, but the windowing code is complicated and I can't be sure. I do believe it's a correct change, in that it simplifies behavior (since we don't actually *need* to unfullscreen on hide or minimize) and improves the behavior on Windows (since previews of minimized windows will have the correct size).

I'd want to avoid any work-around that makes the behavior more complicated or does something different between Windows and Wine.

**Comment by madewokherd** [comment]
Seems like Wine is at fault and not the WM, so I filed a Wine bug here: https://bugs.winehq.org/show_bug.cgi?id=52532

**Comment by icculus** [comment]
@madewokherd Is there any indication the patch mentioned at the winehq bug report is going into Wine (or has it already and the bug is just still open)? I'd like to close this issue if we're confident there's a fix in place elsewhere.

**Comment by madewokherd** [comment]
No one has reviewed it, so no. :(

**Comment by icculus** [comment]
> No one has reviewed it, so no. :(

Is there someone I can nudge on your behalf to review it? I'm happy to do so.

**Comment by madewokherd** [comment]
I asked internally in CW, and it seems Remi has been working on some related things. He also found, to my confusion, that this particular bug doesn't happen in Proton but does happen in Winehq (even though the bug was originally found in testing Proton). I haven't been able to find any Proton change that's likely to have fixed it.

The concern in this area tends to be that changing it is likely to break other things, and without test cases for the things that might break, I have no way to address that. (In this case, I'm selectively disabling code that prevents Wine from making state changes in response to certain events. Maybe that prevents a loop in some application's interaction with an X11 window manager, and I'd break it. Or maybe nothing does that with fullscreen in particular and it's fine. I have no way of knowing because I don't have test cases.)

**Comment by slouken** [comment]
I took a look at this, and it generally seems like an improvement, and the window contents are correct now, where they weren't before. I went ahead and added this for 2.0.22.
